[WiP] ECN3

olantwin opened the pull request:
Please find a work-in-progress version of SHiP@ECN3.

Attached an event display of the setup as it is right now.

Left to do:

- [ ] Scale the detectors (Iaroslava is looking into this)
- [ ] Replace the SND
- [ ] Update the target complex
- [ ] Add and adjust the trenches in ECN3 (I'll try to finish these today)
- [ ] Test, test, test!

![Event display](https://user-images.githubusercontent.com/3578823/172608896-37df2dfd-58cd-4ebf-8653-a7bc2e997d5d.png)

ThomasRuf commented:
while doing all this changes to FairShip, it is propbably better to stop the automatic forwarding to sndsw.

about ClassImp, why is it removed? "The ClassImp and ClassDef macros add members to the class that provide [Run-Time Type Information](https://en.wikipedia.org/wiki/Run-time_type_information) and allow the class to be written to root files."

olantwin commented:
> while doing all this changes to FairShip, it is propbably better to stop the automatic forwarding to sndsw.

Yes, that's probably a good idea.

> about ClassImp, why is it removed? "The ClassImp and ClassDef macros add members to the class that provide [Run-Time Type Information](https://en.wikipedia.org/wiki/Run-time_type_information) and allow the class to be written to root files."

I'll move this into a separate pull request. ClassDef is still necessary, ClassImp isn't (as far as I understand it was used in the old ROOT documentation to allow linking to specific parts of the source code). The commit was included here by accident.

olantwin commented:
> while doing all this changes to FairShip, it is propbably better to stop the automatic forwarding to sndsw.

Done in https://github.com/SND-LHC/sndsw/commit/65d04aba3c3081611a2312827d9b8ffa5fa57210

olantwin commented:
Update: Removed old muon shield options, which allowed doing some cleanup (probably more cleanup possible now). This could be moved to a different pull request, if desirable.

olantwin commented:
Overlaps:

```
Info in <TGeoNodeMatrix::CheckOverlaps>: Checking overlaps for cave and daughters within 0.1
TGeoMatrix::dtor:0: RuntimeWarning: Registered matrix  was removed             
TGeoMatrix::dtor:0: RuntimeWarning: Registered matrix Identity was removed
TGeoMatrix::dtor:0: RuntimeWarning: Registered matrix  was removed          
TGeoMatrix::dtor:0: RuntimeWarning: Registered matrix  was removed
Check overlaps:      [==========] 8367608 [100.00 %]           05:15           
Info in <TGeoNodeMatrix::CheckOverlaps>: Number of illegal overlaps/extrusions : 33

=== Overlaps for FAIRGeom ===
 = Overlap ov00000: cave extruded by: cave/TCC8_1 ovlp=2584
 = Overlap ov00001: cave/MuonShieldArea_1/CoatWall_1 overlapping cave/TCC8_1 ovlp=879.9
 = Overlap ov00002: cave/MuonShieldArea_1/CoatVol_1 overlapping cave/TCC8_1 ovlp=879.9
 = Overlap ov00003: cave/MuonShieldArea_1/AbsorberVol_1 overlapping cave/TCC8_1 ovlp=220
 = Overlap ov00004: cave/ECN3_1 overlapping cave/Tr4_4 ovlp=120.063
 = Overlap ov00005: cave/ECN3_1 overlapping cave/Tr3_3 ovlp=120.063
 = Overlap ov00006: cave/ECN3_1 overlapping cave/Tr1_1 ovlp=119.889
 = Overlap ov00007: cave/ECN3_1 overlapping cave/Tr2_2 ovlp=119.889
 = Overlap ov00008: cave/ECN3_1 overlapping cave/SHiPMagnet_1/magyoke_1 ovlp=118
 = Overlap ov00009: cave/TargetArea_1/moreShieldingTopBot_2 overlapping cave/TCC8_1 ovlp=105
 = Overlap ov00010: cave/ECN3_1 overlapping cave/Hcal_1 ovlp=100
 = Overlap ov00011: cave/ECN3_1 overlapping cave/SHiPMagnet_1/MCoil3_1 ovlp=93
 = Overlap ov00012: cave/ECN3_1 overlapping cave/MuonDetector_1/muonshield_1 ovlp=70
 = Overlap ov00013: cave/ECN3_1 overlapping cave/MuonDetector_1/muondet3_1 ovlp=70
 = Overlap ov00014: cave/ECN3_1 overlapping cave/MuonDetector_1/muonfilter_2 ovlp=70
 = Overlap ov00015: cave/ECN3_1 overlapping cave/MuonDetector_1/muondet2_1 ovlp=70
 = Overlap ov00016: cave/ECN3_1 overlapping cave/MuonDetector_1/muonfilter_1 ovlp=70
 = Overlap ov00017: cave/ECN3_1 overlapping cave/MuonDetector_1/muondet0_1 ovlp=70
 = Overlap ov00018: cave/ECN3_1 overlapping cave/MuonDetector_1/muonfilter_0 ovlp=70
 = Overlap ov00019: cave/ECN3_1 overlapping cave/MuonDetector_1/muondet1_1 ovlp=70
 = Overlap ov00020: cave/ECN3_1 overlapping cave/tTauNuDet_1/PillarVol_1 ovlp=49.99
 = Overlap ov00021: cave/ECN3_1 overlapping cave/tTauNuDet_1/PillarVol_4 ovlp=49.99
 = Overlap ov00022: cave/ECN3_1 overlapping cave/tTauNuDet_1/PillarVol_3 ovlp=49.99
 = Overlap ov00023: cave/ECN3_1 overlapping cave/tTauNuDet_1/PillarVol_2 ovlp=49.99
 = Overlap ov00024: cave/MuonShieldArea_1/Magn2_MagBotLeft_1 overlapping cave/TCC8_1 ovlp=30
 = Overlap ov00025: cave/MuonShieldArea_1/Magn2_MagRetL_1 overlapping cave/TCC8_1 ovlp=29.9
 = Overlap ov00026: cave/MuonShieldArea_1/Magn2_MiddleMagL_1 overlapping cave/TCC8_1 ovlp=29.8544
 = Overlap ov00027: cave/ECN3_1 overlapping cave/SHiPMagnet_1/CV_2 ovlp=8
 = Overlap ov00028: cave/ECN3_1 overlapping cave/SHiPMagnet_1/CV_1 ovlp=8
 = Overlap ov00029: cave/ECN3_1 overlapping cave/SHiPMagnet_1/CV_4 ovlp=8
 = Overlap ov00030: cave/ECN3_1 overlapping cave/SHiPMagnet_1/CV_3 ovlp=8
 = Overlap ov00031: cave/MuonShieldArea_1/MagnAbsorb2_MagRetL_1 overlapping cave/TCC8_1 ovlp=2
 = Overlap ov00032: cave/MuonShieldArea_1/MagnAbsorb2_MagBotLeft_1 overlapping cave/TCC8_1 ovlp=2
TGeoMatrix::dtor:0: RuntimeWarning: Registered matrix Identity was removed
Info in <TGeoNodeMatrix::CheckOverlaps>: Checking overlaps for TargetArea and daughters within 0.0001
Check overlaps:      [==========]   40 [100.00 %]           00:00           
Info in <TGeoNodeMatrix::CheckOverlaps>: Number of illegal overlaps/extrusions : 0

Info in <TGeoNodeMatrix::CheckOverlaps>: Checking overlaps for MuonShieldArea and daughters within 0.0001
Check overlaps:      [==========]   68 [100.00 %]           
Info in <TGeoNodeMatrix::CheckOverlaps>: Number of illegal overlaps/extrusions : 0

Info in <TGeoNodeMatrix::CheckOverlaps>: Checking overlaps for TCC8 and daughters within 0.0001
Check overlaps:      [==========]    1 [100.00 %]           00:00           
Info in <TGeoNodeMatrix::CheckOverlaps>: Number of illegal overlaps/extrusions : 0

Info in <TGeoNodeMatrix::CheckOverlaps>: Checking overlaps for ECN3 and daughters within 0.0001
Check overlaps:      [==========]    1 [100.00 %]           
Info in <TGeoNodeMatrix::CheckOverlaps>: Number of illegal overlaps/extrusions : 0

Info in <TGeoNodeMatrix::CheckOverlaps>: Checking overlaps for SHiPMagnet and daughters within 0.0001
Check overlaps:      [==========]   10 [100.00 %]           
Info in <TGeoNodeMatrix::CheckOverlaps>: Number of illegal overlaps/extrusions : 0

Info in <TGeoNodeMatrix::CheckOverlaps>: Checking overlaps for DecayVolume and daughters within 0.0001
TGeoMatrix::dtor:0: RuntimeWarning: Registered matrix  was removed          
TGeoMatrix::dtor:0: RuntimeWarning: Registered matrix  was removed
Check overlaps:      [==========] 4584 [100.00 %]           
Info in <TGeoNodeMatrix::CheckOverlaps>: Number of illegal overlaps/extrusions : 0

Info in <TGeoNodeMatrix::CheckOverlaps>: Checking overlaps for MagVolume and daughters within 0.0001
Check overlaps:      [==========]    2 [100.00 %]           00:00           
Info in <TGeoNodeMatrix::CheckOverlaps>: Number of illegal overlaps/extrusions : 0

Info in <TGeoNodeMatrix::CheckOverlaps>: Checking overlaps for tTauNuDet and daughters within 0.0001
Check overlaps:      [==========] 9929 [100.00 %]           00:00           
Info in <TGeoNodeMatrix::CheckOverlaps>: Number of illegal overlaps/extrusions : 0

Info in <TGeoNodeMatrix::CheckOverlaps>: Checking overlaps for Tr1 and daughters within 0.0001
Check overlaps:      [==========] 13277 [100.00 %]           0:16            
Info in <TGeoNodeMatrix::CheckOverlaps>: Number of illegal overlaps/extrusions : 0

Info in <TGeoNodeMatrix::CheckOverlaps>: Checking overlaps for Tr2 and daughters within 0.0001
Check overlaps:      [==========] 13277 [100.00 %]           0:17            
Info in <TGeoNodeMatrix::CheckOverlaps>: Number of illegal overlaps/extrusions : 0

Info in <TGeoNodeMatrix::CheckOverlaps>: Checking overlaps for Tr3 and daughters within 0.0001
Check overlaps:      [==========] 13277 [100.00 %]           0:17            
Info in <TGeoNodeMatrix::CheckOverlaps>: Number of illegal overlaps/extrusions : 0

Info in <TGeoNodeMatrix::CheckOverlaps>: Checking overlaps for Tr4 and daughters within 0.0001
Check overlaps:      [==========] 13277 [100.00 %]           0:17            
Info in <TGeoNodeMatrix::CheckOverlaps>: Number of illegal overlaps/extrusions : 0

Info in <TGeoNodeMatrix::CheckOverlaps>: Checking overlaps for Ecal and daughters within 0.0001
Check overlaps:      [==========] 8297606 [100.00 %]           00:00           
Info in <TGeoNodeMatrix::CheckOverlaps>: Number of illegal overlaps/extrusions : 0

Info in <TGeoNodeMatrix::CheckOverlaps>: Checking overlaps for Hcal and daughters within 0.0001
Check overlaps:      [==========]    1 [100.00 %]           00:00           
Info in <TGeoNodeMatrix::CheckOverlaps>: Number of illegal overlaps/extrusions : 0

Info in <TGeoNodeMatrix::CheckOverlaps>: Checking overlaps for MuonDetector and daughters within 0.0001
Check overlaps:      [==========]    9 [100.00 %]           00:00           
Info in <TGeoNodeMatrix::CheckOverlaps>: Number of illegal overlaps/extrusions : 0

Info in <TGeoNodeMatrix::CheckOverlaps>: Checking overlaps for Upstream_Tagger and daughters within 0.0001
Check overlaps:      [==========] 1701 [100.00 %]           
Info in <TGeoNodeMatrix::CheckOverlaps>: Number of illegal overlaps/extrusions : 0

Info in <TGeoNodeMatrix::CheckOverlaps>: Checking overlaps for Timing Detector and daughters within 0.0001
Check overlaps:      [==========]  547 [100.00 %]           00:00           
Info in <TGeoNodeMatrix::CheckOverlaps>: Number of illegal overlaps/extrusions : 0
```

I'll start trying to fix these.

olantwin commented on geometry/geometry_config.py:
```diff
@@ -389,7 +389,7 @@
               c.muShield.dZ7 + c.muShield.dZ8
         ) + c.muShield.LE
-        c.muShield.z = -(c.decayVolume.length + c.muShield.length) / 2.
+        c.muShield.z = -(c.decayVolume.length + c.muShield.length) / 2. + 1.6*u.m
```
Why is this change needed?

yabezsh replied:
This is to move MuonShield closer to the DecayVolume (just for a time being) . The space provided for the TauNu detector was larger than what we anticipate now (5 m). This will probably change with a combi MuonShield.

olantwin replied:
Ah, I see.

olantwin commented:
@yabezsh With the new SND, which overlaps remain?

yabezsh commented:
There is a bunch of overlaps with a cave still there:
 = Overlap ov00000: cave extruded by: cave/TCC8_1 ovlp=2424
 = Overlap ov00013: cave/MuonShieldArea_1/AbsorberVol_1 overlapping cave/TCC8_1 ovlp=220
 = Overlap ov00014: cave/TargetArea_1/moreShieldingTopBot_2 overlapping cave/TCC8_1 ovlp=105
 = Overlap ov00015: cave/MuonShieldArea_1/Magn2_MagBotLeft_1 overlapping cave/TCC8_1 ovlp=30
 = Overlap ov00016: cave/MuonShieldArea_1/Magn2_MagRetL_1 overlapping cave/TCC8_1 ovlp=29.9
 = Overlap ov00017: cave/MuonShieldArea_1/Magn2_MiddleMagL_1 overlapping cave/TCC8_1 ovlp=29.8544
 = Overlap ov00018: cave/MuonShieldArea_1/MagnAbsorb2_MagRetL_1 overlapping cave/TCC8_1 ovlp=2
 = Overlap ov00019: cave/MuonShieldArea_1/MagnAbsorb2_MagBotLeft_1 overlapping cave/TCC8_1 ovlp=2
 = Overlap ov00020: cave/ECN3_1 overlapping cave/SHiPMagnet_1/magyoke_1 ovlp=1


And a set of overlaps within UpstreamTagger. I will investiate from where they are coming.

olantwin commented:
@yabezsh I rebased and fixed the issue with the hadron absorber glitching. I think something must have gone wrong in one of the merges you did, as the hadron absorber and ECN3/TCC8 were different in the version you pushed and my local version.

Please try to rebase instead of merging, I had to deal with many duplicate commits when rebasing, which caused a bit of a headache.

yabezsh commented:
@olantwin
that's a bit weird with an ECN3/TCC8, since I did not touch that part. Anyway, thanks for fixing it! I think we have a more or less fully scaled geometry now?

As for rebasing: sure, I'll do that. I used rebase instead of merge for the last commits (after you mentioned it). I think that was for the first batch of commits. I actually preferred merge in the beginning because it's easier to track. And the branch is not really polluted with only two of us.

olantwin commented:
@yabezsh 
I'm currently updating the hadron absorber and will run another overlap check. But yeah, I think we should be OK then.

The ROOT file for the combi muon shield is on EOS?

yabezsh commented:
> @yabezsh I'm currently updating the hadron absorber and will run another overlap check. But yeah, I think we should be OK then.
> 
> The ROOT file for the combi muon shield is on EOS?

@olantwin  yeap, Eduard sent me this path :  /eos/experiment/ship/user/edursov/opt_combi_fixed/0/(geofile_full.conical.MuonBack-TGeant4.root and combi.root)

olantwin commented:
@yabezsh Do you intentionally leave this much space between SND and muon shield? I think the spacing/length might need to be adjusted.

![Capture d’écran de 2022-07-05 11-44-47](https://user-images.githubusercontent.com/3578823/177300476-6abcbea2-eee3-4bb1-be80-06225d6e9abb.png)

olantwin commented:
There seem to be a lot of overlaps when I use the combi shield. Not entirely sure why, but something messes up the position of the target...

yabezsh commented:
> @yabezsh Do you intentionally leave this much space between SND and muon shield? I think the spacing/length might need to be adjusted.

@olantwin there should not be this gap. I adjusted it (in the Z position of the muon shield), but that was not for the combi shield geometry. I'll correct it. But it should not have had any effect on the target region.

olantwin commented:
It seems like something in the muon shield option 8 might be broken. I'll have a look whether I find something.

olantwin commented:
@yabezsh Shall we try to finish this this week?

For the target-overlaps and length, we should probably check that this bug doesn't exist in the current master branch.

olantwin commented:
With master and the combi-shield, everything looks OK.

olantwin commented on geometry/geometry_config.py:
```diff
@@ -389,46 +389,10 @@
-     c.muShield.length = 2*(c.muShield.dZ1+c.muShield.dZ2+c.muShield.dZ3+c.muShield.dZ4+c.muShield.dZ5+c.muShield.dZ6
-                         +c.muShield.dZ7+c.muShield.dZ8 ) + c.muShield.LE  # leave some space for nu-tau 
-     c.muShield.z  =  -c.decayVolume.length/2.-c.muShield.length/2.
+        c.muShield.z = -(c.decayVolume.length + c.muShield.length) / 2. + 1.6*u.m
```
@yabezsh  This line seems to be the culprit. I don't understand why this only seems to be a problem for the combi-shield (or muon shield 8, but if I disable it, the target is in the right spot.

Can't run an overlap check until this evening unfortunately.

olantwin replied:
@yabezsh Reverting the change to this line also fixes the overlaps with between target and absorber.

yabezsh replied:
Interesting... it seems that if we don't use option 8 of muShield, this line just moves Shield and everything behind it closer to the SND area. However, with option 8 it doesn't move closer to the SND area at all, instead the target is pushed a bit inside Shield. It seems that the position variables for the two shield options are related in different ways

olantwin replied:
Yes, and I really can't figure out why. I feel like I'm missing something really obvious somewhere...

ThomasRuf commented:
Just want to mention that the positioning of the muon shield/target was done such, that the vacuum tank does not had to move all the time the muon shield had been optimized. Therefore, it was necessary to calculate first the total length of target/muon shield/ SND, to know the position of the start of the target.

olantwin commented:
Hi Thomas,

The configurable muon shield should follow this idea. Based on its parameters its length is calculated in geometry_config. This definitely used to work. I'll check whether a bug might have been introduced in the meantime.

olantwin commented:
OK, I think I understand the issue now.

I'll fix the issue later before Monday.

@yabezsh I also found the proper location to adjust the length of the SND: `c.muonShield.LE`, currently set to 10m.

olantwin commented:
This should now be ready for testing and review.

Minor overlaps in the UVT, and between cavern and ShipMagnet and target shielding remain (see https://github.com/olantwin/FairShip/issues/52), and will be hopefully be fixed soon.

olantwin commented:
EventDisplay of the geometry with the combi shield

![Capture d’écran de 2022-07-09 12-20-29](https://user-images.githubusercontent.com/3578823/178101635-f47ba577-a8bf-49f6-82b7-ca24cbcacee7.png)

olantwin commented:
Closeup of SND after fix.

![Capture d’écran de 2022-07-05 11-44-47](https://user-images.githubusercontent.com/3578823/178102761-7558fe99-204a-4eb2-acd9-d55671e14a41.png)

olantwin commented on UpstreamTagger/UpstreamTagger.cxx:
```diff
@@ -379,7 +379,7 @@ void UpstreamTagger::ConstructGeometry()
 
   InitMedium("vacuum");  
   TGeoMedium *Vacuum_box =gGeoManager->GetMedium("vacuum");
-  double xbox=233.4; double ybox = 113; double zbox = 1.3503;
```
@yabezsh Did you ever understand how these numbers are calculated?

yabezsh replied:
I do not remember exactly what these values are, but if you plot them, it becomes clear. Essentially, they are different blocks, and you can tell by their position that they should be shorter, etc. I have no idea if the numbers are from real size blocks. We need someone who is familiar with UT

olantwin replied:
For now I scaled the the numbers naively for #420, and the result looks more or less OK. Let's see what happens once we check for overlaps...
Scaling of the other parts seems more difficult right now. Things didn't quite scale as I expected.
